libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h

```diff
@@ -0,0 +1,213 @@
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#if !defined(_CCCL_CUDA_COMPILER_NVCC) && !defined(_CCCL_CUDA_COMPILER_NVHPC)
+#  include <cuda_runtime_api.h>
+#endif // !_CCCL_CUDA_COMPILER_NVCC && !_CCCL_CUDA_COMPILER_NVHPC
+
+#include <cuda/__memory_resource/get_property.h>
+#include <cuda/__memory_resource/properties.h>
+#include <cuda/__memory_resource/resource_ref.h>
+#include <cuda/__memory_resource/resource.h>
+
+#if _CCCL_STD_VER >= 2014
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA_MR
+
+/**
+ * @brief `cuda_memory_resource` uses cudaMalloc / cudaFree for allocation/deallocation.
+ */
+struct cuda_memory_resource
+{
+  /**
+   * @brief Allocate device memory of size at least \p __bytes.
+   * @param __bytes The size in bytes of the allocation.
+   * @param __alignment The requested alignment of the allocation. Is ignored!
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* allocate(const size_t __bytes, const size_t __alignment) const
+  {
+    _LIBCUDACXX_ASSERT(__alignment <= 256 && (256 % __alignment == 0),
+                       "cuda_memory_resource::allocate invalid alignment");
+    return allocate(__bytes);
+  }
+
+  /**
+   * @brief Allocate device memory of size at least \p __bytes.
+   * @param __bytes The size in bytes of the allocation.
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* allocate(const size_t __bytes) const
+  {
+    void* __ptr{nullptr};
+    const ::cudaError_t __status = ::cudaMalloc(&__ptr, __bytes);
+    switch (__status)
+    {
+      case ::cudaSuccess:
+        return __ptr;
+      default:
+        ::cudaGetLastError(); // Clear CUDA error state
+#  ifndef _LIBCUDACXX_NO_EXCEPTIONS
+        throw cuda::cuda_error{__status, "Failed to allocate memory with cudaMalloc."};
+#  else
+        _LIBCUDACXX_UNREACHABLE();
+#  endif
+    }
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by \p __ptr.
+   * @param __ptr Pointer to be deallocated
+   * @param __bytes The size in bytes of the allocation.
+   * @param __alignment The alignment that was passed to the `allocate` call that returned \p __ptr. Is ignored!
+   */
+  void deallocate(void* __ptr, const size_t __bytes, const size_t __alignment) const
+  {
+    _LIBCUDACXX_ASSERT(__alignment <= 256 && (256 % __alignment == 0),
+                       "cuda_memory_resource::deallocate invalid alignment");
+    deallocate(__ptr, __bytes);
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by \p __ptr.
+   * @param __ptr Pointer to be deallocated
+   * @param __bytes The size in bytes of the allocation.
+   */
+  void deallocate(void* __ptr, size_t) const
+  {
+    const ::cudaError_t __status = ::cudaFree(__ptr);
+    (void)__status;
+    _LIBCUDACXX_ASSERT(__status == cudaSuccess, "cuda_memory_resource::deallocate failed");
+  }
+
+  /**
+   * @brief Equality comparison operator between two cuda_memory_resource's
```

Comment:
suggestion: Apostrophe for possessives and contractions, not for plural. It's awkward when you want doxygen linking, but perhaps just simplify to use "another" rather than "two", and no need to say this is an operator. :)

```suggestion
   * @brief Equality comparison with another cuda_memory_resource
```